Tests/PromiseProgressTests.cs

```diff
@@ -0,0 +1,188 @@
+
+            var expectedSteps = new float[] { 0.1f, 0.2f, 0.6f, 1f };
+            var currentProgress = 0f;
+            int currentStep = 0;
+
+            promiseC.
+                Progress(v =>
+                {
+                    Assert.InRange(currentStep, 0, expectedSteps.Length - 1);
+                    Assert.InRange(v - expectedSteps[currentStep], -Math.E, Math.E);
+                    currentProgress = v;
+                    ++currentStep;
+                })
+            ;
+
+            promiseA.
+                Progress(v => promiseC.ReportProgress(v * 0.2f))
+                .Then(() => promiseB)
+                .Progress(v => promiseC.ReportProgress(0.2f + 0.8f * v))
+                .Then(() => promiseC.Resolve())
+                .Catch(ex => promiseC.Reject(ex))
+            ;
+
+            promiseA.ReportProgress(0.5f);
+            promiseA.ReportProgress(1f);
+            promiseA.Resolve();
+            promiseB.ReportProgress(0.5f);
+            promiseB.ReportProgress(1f);
+            promiseB.Resolve();
+
+            Assert.Equal(1f, currentProgress);
+        }
+
+
+        [Fact]
+        public void chain_multiple_promises_reporting_progress()
+        {
+            var promiseA = new Promise();
+            var promiseB = new Promise();
+            var progressA = 0f;
+            var progressB = 0f;
+
+            promiseA
+                .Progress(v => progressA = v)
+                .Then(() => promiseB)
+                .Progress(v => progressB = v)
+                .Done();
+
+            promiseA.ReportProgress(1f);
+            promiseA.Resolve();
+            promiseB.ReportProgress(2f);
+            promiseB.Resolve();
+
+            Assert.Equal(1f, progressA);
+            Assert.Equal(2f, progressB);
+        }
+
+        [Fact]
+        public void exception_is_thrown_for_progress_after_resolve()
+        {
+            var promise = new Promise();
+            promise.Resolve();
+
+            Assert.Throws<ApplicationException>(() => promise.ReportProgress(1f));
+        }
+
+        [Fact]
+        public void exception_is_thrown_for_progress_after_reject()
+        {
+            var promise = new Promise();
+            promise.Reject(new ApplicationException());
+
+            Assert.Throws<ApplicationException>(() => promise.ReportProgress(1f));
+        }
+
+        [Fact]
+        public void all_progress_is_averaged()
+        {
+            var promiseA = new Promise();
+            var promiseB = new Promise();
+            var promiseC = new Promise();
+            var promiseD = new Promise();
+
+            int currentProgress = 0;
+            var expectedProgress = new float[] { 0.25f, 0.50f, 0.75f, 1f };
+
+            Promise.All(new IPromise[] { promiseA, promiseB, promiseC, promiseD })
+                .Progress(progress =>
+                {
+                    Assert.InRange(currentProgress, 0, expectedProgress.Length - 1);
```

Comment:
currentProgress is only set once to 0, so these asserts aren't really testing anything useful.

Reply:
Funnily enough, currentProgress should be incremented on every callback call, so it passed a test when it shouldn't have. I'll check these later.